File: libs/deno-lib/src/api_response.ts
export interface ApiResponseType<T> {
  success: boolean;
  message: string | null;
  data: T | null;
}

export class ApiResponse<T> implements ApiResponseType<T> {
  readonly success: boolean;
  readonly message: string | null;
  readonly data: T | null;

  private constructor(success: boolean, message: string | null, data: T | null) {
    this.success = success;
    this.message = message;
    this.data = data;
  }

  static success<T>(data: T, message?: string): ApiResponse<T> {
    return new ApiResponse(true, message ?? null, data);
  }

  static error<T>(message: string): ApiResponse<T> {
    return new ApiResponse<T>(false, message, null as T | null);
  }

  toJSON(): ApiResponseType<T> {
    return {
      success: this.success,
      message: this.message,
      data: this.data,
    };
  }
}

File: libs/dotnet-lib/ApiResponse.cs
using System.Text.Json.Serialization;

namespace CoinHub.Common;

public class ApiResponse<T> where T : class
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("message")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Message { get; set; }

    [JsonPropertyName("data")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public T? Data { get; set; }

    private ApiResponse(bool success, string? message, T? data)
    {
        Success = success;
        Message = message;
        Data = data;
    }

    public static ApiResponse<T> SuccessResponse(T data) =>
        new(true, null, data);

    public static ApiResponse<T> SuccessWithMessage(string message, T data) =>
        new(true, message, data);

    public static ApiResponse<T> ErrorResponse(string message) =>
        new(false, message, null);
}
